spec/unit/functions/any_resources_of_type_spec.rb

```diff
@@ -0,0 +1,87 @@
+require 'spec_helper'
+require 'puppet/pops'
+require 'puppet/loaders'
+
+describe "the 'any_resources_of_type' function" do
+  after(:all) { Puppet::Pops::Loaders.clear }
+
+  # This loads the function once and makes it easy to call it
+  # It does not matter that it is not bound to the env used later since the function
+  # looks up everything via the scope that is given to it.
+  # The individual tests needs to have a fresh env/catalog set up
+  #
+  let(:loaders) { Puppet::Pops::Loaders.new(Puppet::Node::Environment.create(:testing, [File.expand_path(fixtures('modules'))])) }
+  let(:func) { loaders.private_environment_loader.load(:function, 'any_resources_of_type') }
+
+  before :each do
+    # A fresh environment is needed for each test since tests create resources
+    environment = Puppet::Node::Environment.create(:testing, [File.expand_path(fixtures('modules'))])
+    @node = Puppet::Node.new('yaynode', :environment => environment)
+    @compiler = Puppet::Parser::Compiler.new(@node)
+    @scope = Puppet::Parser::Scope.new(@compiler)
+  end
+
+  def newresource(type, title, parameters = {})
+    resource = Puppet::Resource.new(type, title, :parameters => parameters)
+    @compiler.add_resource(@scope, resource)
+    resource
+  end
+
+  it 'raises if called without the "resource_type" argument' do
+    expect { func.call(@scope) }.to raise_error(ArgumentError, %r{expects between.*got none})
+  end
+
+  it 'allows an optional "parameters" argument' do
+    expect { func.call(@scope, 'title', { :param1 => 'value1' }) }.not_to raise_error
+  end
+
+  it 'raises if the "parameters" argument is not a hash' do
+    expect { func.call(@scope, 'title', 'not_a_hash') }.to raise_error(ArgumentError, %r{parameter 'parameters' expects a Hash value, got String})
+  end
+
+  context 'can determine if a resource' do
+    context 'is defined' do
+      it 'by looking up the type name' do
```

Comment:
[Optional]

I would restructure this to something like:
```ruby
  context 'when resource is defined' do
    it 'finds by type name' do
	...

  context 'when resource is not defined' do
	it 'does not find by type name'
```

Reply:
Makes sense. I'll also add with multiple resources of the same type.